core/modules/replica/DatabaseServicesMySQL.cc

```diff
@@ -212,184 +231,200 @@ void DatabaseServicesMySQL::saveState(QservMgtRequest const& request,
-            0,
-            QservMgtRequest::state2string(request.state()),
-            QservMgtRequest::state2string(request.extendedState()),
-            serverError,
-            performance.c_create_time,
-            performance.c_start_time,
-            performance.w_receive_time,
-            performance.w_start_time,
-            performance.w_finish_time,
-            performance.c_finish_time);
-
-        // Extended state (if any provided by a specific request class) is recorded
-        // in a request-specific table whose name is based on a value of the request's
-        // 'type' parameter.
-
-        std::string extendedTableName = "request_" + request.type();
-        std::transform(extendedTableName.begin(),
-                       extendedTableName.end(),
-                       extendedTableName.begin(),
-                       [] (unsigned char c) {
-                           return std::tolower(c);
-                       });
-
-        std::string const extendedPersistentState = request.extendedPersistentState(_conn);
-        if (not extendedPersistentState.empty()) {
-            LOGS(_log, LOG_LVL_DEBUG, context << "extendedPersistentState: " << extendedPersistentState);
-            _conn->execute("INSERT INTO " + _conn->sqlId(extendedTableName) +
-                           " VALUES " + extendedPersistentState);
-        }
-        _conn->commit ();
-
-    } catch (database::mysql::DuplicateKeyError const&) {
+    // The algorithm will first try the INSERT query into the base table.
+    // If a row with the same primary key (QservMgtRequest id) already exists in the table
+    // then the UPDATE query will be executed.
 
+    try {
         try {
-            _conn->rollback();
-            _conn->begin();
-            _conn->executeSimpleUpdateQuery(
-                "request",
-                _conn->sqlEqual("id",                                           request.id()),
-                std::make_pair( "state",          QservMgtRequest::state2string(request.state())),
-                std::make_pair( "ext_state",      QservMgtRequest::state2string(request.extendedState())),
-                std::make_pair( "server_status",                                serverError),
-                std::make_pair( "c_create_time",  performance.c_create_time),
-                std::make_pair( "c_start_time",   performance.c_start_time),
-                std::make_pair( "w_receive_time", performance.w_receive_time),
-                std::make_pair( "w_start_time",   performance.w_start_time),
-                std::make_pair( "w_finish_time",  performance.w_finish_time),
-                std::make_pair( "c_finish_time",  performance.c_finish_time));
-
-            _conn->commit();
-
-        } catch (database::mysql::Error const& ex) {
-            if (_conn->inTransaction()) _conn->rollback();
-            throw std::runtime_error(context + "failed to save the state, exception: " + ex.what());
+            _conn->execute(
```

Comment:
try try again! ...maybe this could be factored into named helper functions? It might make the logic more clear and the details neatly hidden.

Reply:
Frankly, I don't see any problem with this code. Nested 'try's aren't much different from nested namespace declaration or nested loops. In this particular pattern the inner try catches the key duplicate condition (which is a normal situation) on an attempt to insert a new row and updates a database record if the key already exists. This is a standard MySQL programming paradigm for dealing with this kind of scenarios.
And the outer loop is meant to catch the remaining exceptions originating from any branch of the inner 'try'.
Though, I've noted your concern.